docs/stability-platform-support.md

```diff
@@ -42,3 +42,71 @@ cases for for what `#[no_std]` might entail, so if you're interested in this
 issue](https://github.com/bytecodealliance/wasmtime/issues/new) on the
 `wasmtime` repository to discuss this.
+
+This is a common question we are asked, however, so to provide some more context
+on why Wasmtime is the way it is, here's some responses to frequent points
+raised about `#![no_std]`:
+
+* **What if my platform doesn't have `std`?** - For platforms without support
+  for the Rust standard library the JIT compiler of Wasmtime often won't run on
+  the platform as well. The JIT compiler requires `mmap` (or an equivalent), and
+  presence of `mmap` often implies presence of a libc which means Rust's `std`
+  library works. We're interested in running Wasmtime without a JIT compiler in
+  the future, but that is not implemented at this time. Implementing this will
+  require a lot more work than tagging crates `#![no_std]` as well! The Wasmtime
+  developers are also very interested in supporting as many targets as possible,
+  so if Wasmtime doesn't work on your platform yet we'd love to learn why and
+  what we can do to support that platform, but the conversation here is
+  typically more nuanced than simply making `wasmtime` compile without `std`.
+
+* **Doesn't `#![no_std]` have smaller binary sizes?** - There's a lot of factors
+  that affect binary size in Rust. Compilation options are a huge one but beyond
+  that idioms and libraries linked matter quite a lot as well. Code is not
+  inherently large when using `std` instead of `core`, it's just that often code
+  using `std` has more dependencies (like `std::thread`) which requires code to
+  bind. Code size improvements can be made to code using `std` and `core`
+  equally, and switching to `#![no_std]` is not a silver bullet for compile
+  sizes.
+
+* **The patch to switch to `#![no_std]` is small, why not accept it?** - PRs to
+  switch to `#![no_std]` are often relatively small or don't impact too many
+  parts of the system. There's a lot more to developing a `#![no_std]`
+  WebAssembly runtime than switching a few crates, however. Maintaining a
+  `#![no_std]` library over time has a number of costs associated with it:
+
+  * There needs to be CI to ensure that when compiled with the right flags the
+    library does not actually use the Rust standard library. Currently there is
+    no stable way to do this in Rust, meaning that although a library may be
+    `#![no_std]` at one point in time it's easy to add a dependency that
+    accidentally sneaks in the `std` crate later.
+
+  * Idioms in `#![no_std]` are quite different than normal Rust code. You'll
+    import from different crates (`core` instead of `std`) and data structures
+    have to all be manually imported from `alloc`. These idioms are difficult to
+    learn for newcomers to the project and are not well documented in the
+    ecosystem. This cost of development and maintenance is not unique to
+    Wasmtime but in general affects the `#![no_std]` ecosystem at large,
+    unfortunately.
```

Comment:
I personally feel like `no-std-compat` is a good example of "if it compiles it works, right?" with a big emphasis on the question mark. It's trying to make the ecosystem better but in a way that fundamentally is not possible. The best example of this is the dependency on the `spin` crate to implement `std::sync::Mutex`. A spin lock is basically never what you want unless you're a kernel and can disable interrupts. Otherwise it's attempting to sidestep fundamental questions of "how do I do multithreading or interact with the system scheduler" where you can't really sidestep these questions.

Crates like that also seem to be taking the fundamental stance of "std will never attempt to fix any of these problems, right?" when in fact PRs like https://github.com/rust-lang/rust/pull/74033 will basically fix the issue for us. If all you want is code to compile using `std` as-is the way it's written today, then we should be putting energy behind support in the standard library itself upstream. There really is no reason that `std` can't do what something like `no-std-compat` already does in a way that supports the already existing ecosystem idioms.

Reply:
I agree, I'm just considering the perspective of a user of Wasmtime asking about `no_std` support: `no-std-compat` exists and works today, and doesn't do anything that would get in the way of better solutions.

Reply:
Do you think it's worth explicitly calling this out in the documentation here? Basically adding my comment as a new "what about `no-std-compat`" FAQ?

Reply:
If  rust-lang/rust#74033 means that our options will change in the not too distant future, one option here is to replace the above paragraph with one that links to that, and points out that we'd prefer to avoid the thrash of updating all the files to a temporary solution when a better one is one the way. Then I don't think we'd even need to mention `no-std-compat`.

Reply:
I think we should at least link to `build-std`. However, IIUC it'll be quite some time until that reaches stable, so it'll probably not be a "just do this" answer anytime soon. Given that, perhaps we could also link to `no-std-compat` and point out that people *can* use it if it works for their use case, but that we don't include it for the reasons @alexcrichton stated above?

Reply:
`no-std-compat` by itself isn't sufficient to make Wasmtime run in `no_std`, because at least `mmap`. We may make that avoidable at some point, but that'll take time too. So I suggest just linking to `build-std`, and not `no-std-compat`, and just say that we'll wait until the better solutions arrive.

Also, I think it's reasonable to ask people who want to do `no_std` things and don't want to wait for `build-std` features to stabilize to use nightly Rust for a Rust release cycle or two.

Reply:
One thing I don't really understand though is in what context we're writing this information down. This section is under the question "the patch is small, why not?" and this sub-point is "the idioms are different enough that it's nontrivial to do so". I don't get the impression that people are frequently sending patches with `no-std-compat` being used. Apart from that I'm not really sure how to tweak the words already there.

If y'all want I can write an explicit bullet saying "`no-std-compat` doesn't work", or I can add more words under "my target doesn't have `std`" pointing to `-Zbuild-std` and the PR I mentioned, but neither of those really changes the point that the idioms of `#![no_std]` are different than that of `std` in nontrivial ways.

Also, unrelated to this PR itself, but do y'all think that we should be moving to `#![no_std]` today? I can't quite get a feeling for whether y'all are playing devil's advocate or are instead along the lines of "I think we should do this and these words don't convince me we shouldn't"

Reply:
I agree we shouldn't do `#![no_std]` today. As you mention, until we have a way to avoid calling `mmap`, it wouldn't matter.

The `-Zbuild-std` thing sounds cool. I think we should mention/link to that, and say that it doesn't make sense to take on `no_std` changes with the current tools when better tools are on the horizon.

Reply:
I just left another comment on the first question with a suggestion for how this might be structured. Also agreed that mentioning `no-std-compat` might not be needed.